src/Chainweb/Mempool/Consensus.hs

```diff
@@ -94,6 +66,95 @@ data ReintroducedTxs = ReintroducedTxs
     deriving (Eq, Show, Generic)
     deriving anyclass (ToJSON, NFData)
 
+newtype MempoolException = MempoolConsensusException String
+
+instance Show MempoolException where
+    show (MempoolConsensusException s) = "Error with mempool's consensus processing: " ++ s
+
+instance Exception MempoolException
+
+----------------------------------------------------------------------------------------------------
+mkMempoolConsensus :: PayloadCas cas => MempoolBackend t -> BlockHeaderDb -> Maybe (PayloadDb cas) -> IO (MempoolConsensus t)
+mkMempoolConsensus mempool blockHeaderDb payloadStore = do
+              lastParentRef <- newIORef Nothing :: IO (IORef (Maybe BlockHeader))
+              return MempoolConsensus
+                  { mpcMempool = mempool
+                  , mpcLastNewBlockParent = lastParentRef
+                  , mpcProcessFork = processFork blockHeaderDb payloadStore lastParentRef
+                  }
+
+----------------------------------------------------------------------------------------------------
+processFork :: PayloadCas cas
+                 => BlockHeaderDb
+                 -> Maybe (PayloadDb cas)
+                 -> IORef (Maybe BlockHeader)
+                 -> LogFunction
+                 -> BlockHeader
+                 -> IO (Vector ChainwebTransaction)
+processFork blockHeaderDb payloadStore lastHeaderRef logFun newHeader = do
+    lastHeader <- readIORef lastHeaderRef
```

Comment:
Don't you need to write back to the ref after syncing?

Reply:
No, because I'm updating that 'lastHeader' value at validate time.  Based on when this is called, the the 'newHeader' here is coming from a newBlock request.  

I considered updating the 'lastHeader' at both newBlock and validateBlock time, but that doesnt seem right, given that a "new Block" may likely not win and be validated.

Reply:
I think it should only happen at validate time? New blocks just get sent to front of consensus and are re-validated, right?

Reply:
Yes, Its only at validate time.  And right, new blocks get validated in the same way blocks from other nodes are validated